src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,973 @@
+  (let 
+    ((current-page (var-get attachments-inv-page-cursor))
+    (current-index (var-get attachments-inv-index-cursor)))
+    (let 
+      ((next-page (if (is-eq (+ current-index u1) ATTACHMENTS_INV_PAGE_SIZE)
+        (+ current-page u1)
+        current-page))
+      (next-index (mod (+ current-index u1) ATTACHMENTS_INV_PAGE_SIZE)))
+      ;; Emit event used as a system hinter
+      (print {
+        attachment: {
+          hash: zonefile-hash,
+          page-index: current-page,
+          position-in-page: current-index,
+          metadata: {
+            name: name,
+            namespace: namespace,
+            tx-sender: tx-sender
+          }
+        }})
+      ;; Update attachments-inv
+      (map-set attachments-inv
+        ((page current-page) (index current-index))
+        ((content-hash zonefile-hash)))
+      ;; Update cursors
+      (var-set attachments-inv-page-cursor next-page)
+      (var-set attachments-inv-index-cursor next-index)
+      (map-set name-properties
+        ((namespace namespace) (name name))
+        ((registered-at registered-at)
+          (imported-at imported-at)
+          (revoked-at revoked-at)
+          (zonefile-hash zonefile-hash))))))
+
+(define-read-only (get-attachments-inv-info)
+  (ok { 
+    pages-count: (+ (var-get attachments-inv-page-cursor) u1),
+    last-page-len: (var-get attachments-inv-index-cursor),
+    page-size: ATTACHMENTS_INV_PAGE_SIZE
+  }))
+
+;;;; NAMESPACES
+;; NAMESPACE_PREORDER
+;; This step registers the salted hash of the namespace with BNS nodes, and burns the requisite amount of cryptocurrency.
+;; Additionally, this step proves to the BNS nodes that user has honored the BNS consensus rules by including a recent
+;; consensus hash in the transaction.
+;; Returns pre-order's expiration date (in blocks).
+(define-public (namespace-preorder (hashed-salted-namespace (buff 20))
+                                   (stx-to-burn uint))
+  (let 
+    ((former-preorder 
+      (map-get? namespace-preorders ((hashed-salted-namespace hashed-salted-namespace) (buyer contract-caller)))))
+    ;; Ensure eventual former pre-order expired 
+    (asserts! 
+      (if (is-none former-preorder)
+        true
+        (>= block-height (+ NAMESPACE_PREORDER_CLAIMABILITY_TTL
+                            (unwrap! (get created-at former-preorder) (err ERR_PANIC)))))
+      (err ERR_NAMESPACE_PREORDER_ALREADY_EXISTS))
+          (asserts! (> stx-to-burn u0) (err ERR_NAMESPACE_STX_BURNT_INSUFFICIENT))
+    ;; Ensure that the hashed namespace is 20 bytes long
+    (asserts! (is-eq (len hashed-salted-namespace) u20) (err ERR_NAMESPACE_HASH_MALFORMED))
+    ;; Ensure that user will be burning a positive amount of tokens
+    (asserts! (> stx-to-burn u0) (err ERR_NAMESPACE_STX_BURNT_INSUFFICIENT))
+    ;; Burn the tokens
+    (unwrap! (stx-transfer? stx-to-burn contract-caller BURN_ADDRESS) (err ERR_INSUFFICIENT_FUNDS))
```

Comment:
Shouldn't this be `stx-burn?`, or is that dropped?

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/0bdf51f9f85cb45ddf8340b45838c7d5ef1c6984